### ISSUE TITLE
process_data_at_eof's 1024 bytes offset is not enough

### ISSUE THREAD

**Issue by fredericoschardong** [MatthiasValvekens/pyHanko#173]
**Describe the bug**
Dear Matthias,

On function [`process_data_at_eof`](https://github.com/MatthiasValvekens/pyHanko/blob/master/pyhanko/pdf_utils/reader.py#L101), there appears to be an arbitrary 1024-byte offset:

```
# offset of last 1024 bytes of stream
last_1k = stream.tell() - 1024 + 1
```

There are some PDFs whose `%%EOF` is further back from the end. For example, the `%%EOF` of this file [17a5ac7f-991f-45d2-900e-35997c3d22a7.pdf](https://github.com/MatthiasValvekens/pyHanko/files/9844658/17a5ac7f-991f-45d2-900e-35997c3d22a7.pdf) is 1037 bytes away from the end.

**To Reproduce**
Download the file then run:
```
openssl req -x509 -sha256 -nodes -days 365 -newkey rsa:2048 -keyout privateKey.key -out certificate.crt
pyhanko sign addsig --field Sig1 pemder --no-pass --key privateKey.key --cert certificate.crt 17a5ac7f-991f-45d2-900e-35997c3d22a7.pdf out.pdf
```

**Screenshots**
```
pyhanko.cli - ERROR - Failed to read PDF file: EOF marker not found
Error: Failed to read PDF file: EOF marker not found
```

**Environment (please complete the following information):**
 - Ubuntu 20.04, Python 3.10.6, pyHanko 0.15.0

**Comment by MatthiasValvekens** [comment]
Hi @fredericoschardong,


That bound was inherited from PyPDF2 if I recall correctly, and as far as I'm concerned, it's more about making allowances for discrepancies in the precise positioning of the startxref directive than about tolerating `%%EOF` not being at the end of the stream. If anything, the number could be made significantly smaller: the specification is quite explicit about no data being permitted after `%%EOF` (other than perhaps an EOL sequence).

Nonetheless, the number 1024 is (kinda) industry standard. The older (pre-ISO) Adobe specs used to include these words in the implementation-specific section:

> 18. Acrobat viewers require only that the %%EOF marker appear somewhere within the last 1024 bytes of the file.

Other libraries like iText [also follow this convention](https://github.com/itext/itext7/blob/68055cba7740381c718ae965f0786060ba64e919/io/src/main/java/com/itextpdf/io/source/PdfTokenizer.java#L248-L262).


Besides, pyHanko assumes that the `%%EOF` rule is satisfied for the purposes of computing signature bounds and whatnot, so increasing the limit won't really help you in this particular case. If anything, pyHanko should probably enforce the `%%EOF` rule by default, since it will silently create (potentially) invalid signatures otherwise.

-----------------------------------------------

So, long story short: I'm not very keen on giving _more_ leeway to nonconforming PDFs, especially in a signing library. I would be willing to accept a PR to increase tolerances, with some conditions:

 - The additional allowances are not applied in strict mode, and a check is added to strict mode to make sure the `%%EOF` really is at the end of the file.
 - The signing/validation code is also amended to work with the "logical" `%%EOF` rather than the "physical" one.
 - All changes are well-tested with example documents for both strict and nonstrict modes to ensure there are no undesirable side effects.

If your project needs to support files with garbage data at the end, then I'd personally prefer it if you'd just preprocess them to strip out said garbage data :).
